src/binary/cursor.rs

```diff
@@ -523,6 +525,96 @@ impl<R: IonDataSource> Cursor for BinaryIonCursor<R> {
         self.clob_ref_map(|c| c.into())
     }
 
+    fn read_timestamp(&mut self) -> IonResult<Option<Timestamp>> {
```

Comment:
This method feels a bit long to me. I could make it a bit more concise by either:
* Defining some traits for the various Timestamp builder API types to allow `build()`, `build_at_offset()`, and `build_at_unknown_offset()` to be called generically. This would let me write a couple of helper methods.
* Defining a macro for the 
```rust
if self.finished_reading_value() {
  let timestamp = // ...
  return Ok(Some(timestamp));
}
```
boilerplate repeated for each field.

The main drawback is that this is the only function where those artifacts would be very useful. I'm not sure how much net benefit we'd get for the investment. Open to other opinions, though!